interceptor/src/stats/interceptor.rs

```diff
@@ -1,58 +1,217 @@
+}
+
 pub struct StatsInterceptor {
+    // Wrapped RTP streams
     recv_streams: Mutex<HashMap<u32, Arc<RTPReadRecorder>>>,
     send_streams: Mutex<HashMap<u32, Arc<RTPWriteRecorder>>>,
+
+    tx: mpsc::Sender<Message>,
+
     id: String,
+    now_gen: Option<Arc<dyn Fn() -> SystemTime + Send + Sync>>,
 }
 
 impl StatsInterceptor {
     pub fn new(id: String) -> Self {
+        let (tx, rx) = mpsc::channel(100);
+
+        tokio::spawn(run_stats_reducer(rx));
+
         Self {
             id,
             recv_streams: Default::default(),
             send_streams: Default::default(),
+            tx,
+            now_gen: None,
         }
     }
 
-    pub fn recv_stats_reader(&self, ssrc: u32) -> Option<RTPStatsReader> {
-        self.recv_stats_readers([ssrc].iter().copied())
-            .into_iter()
-            .next()
+    fn with_time_gen<F>(id: String, now_gen: F) -> Self
+    where
+        F: Fn() -> SystemTime + Send + Sync + 'static,
+    {
+        let (tx, rx) = mpsc::channel(100);
+        tokio::spawn(run_stats_reducer(rx));
+
+        Self {
+            id,
+            recv_streams: Default::default(),
+            send_streams: Default::default(),
+            tx,
+            now_gen: Some(Arc::new(now_gen)),
+        }
     }
 
-    pub fn recv_stats_readers(&self, ssrcs: impl Iterator<Item = u32>) -> Vec<RTPStatsReader> {
-        let lock = self.recv_streams.lock();
+    pub async fn fetch_stats(&self, ssrcs: Vec<u32>) -> Vec<Option<StatsSnapshot>> {
+        let (tx, rx) = oneshot::channel();
 
-        ssrcs
-            .filter_map(|ssrc| lock.get(&ssrc).map(|r| r.reader()))
-            .collect()
-    }
+        if let Err(e) = self
+            .tx
+            .send(Message::RequestSnapshot { ssrcs, chan: tx })
+            .await
+        {
+            log::debug!(
+                "Failed to fetch RTP stream stats from stats task with error: {}",
+                e
+            );
 
-    pub fn send_stats_reader(&self, ssrc: u32) -> Option<RTPStatsReader> {
-        self.send_stats_readers([ssrc].iter().copied())
-            .into_iter()
-            .next()
-    }
+            return vec![];
+        }
 
-    pub fn send_stats_readers(&self, ssrcs: impl Iterator<Item = u32>) -> Vec<RTPStatsReader> {
-        let lock = self.send_streams.lock();
+        rx.await.unwrap_or_default()
+    }
+}
 
-        ssrcs
-            .filter_map(|ssrc| lock.get(&ssrc).map(|r| r.reader()))
-            .collect()
+async fn run_stats_reducer(mut rx: mpsc::Receiver<Message>) {
+    let mut ssrc_stats: StatsContainer = Default::default();
+    let mut cleanup_ticker = tokio::time::interval(Duration::from_secs(10));
+
+    loop {
+        tokio::select! {
+            maybe_msg = rx.recv() => {
+                let msg = match maybe_msg {
+                    Some(m) => m,
+                    None => break,
+                };
+                match msg {
+                    Message::StatUpdate { ssrc, update } => {
```

Comment:
Maybe we move this whole thing to a function since formatting and highlighting breaks within `tokio::select!` because it's a macro